lib/environment/permissions.js

```diff
@@ -8,25 +8,39 @@ function filterPerms(perms) {
 	return { permissions, origins };
 }
 
-const rejectListeners = [];
-
-export function onReject(callback) {
-	rejectListeners.push(callback);
-}
-
 /**
  * @param {...string} perms Optional Chrome permissions to request.
  * @returns {Promise<void, *>} Resolves if the permissions are granted, rejects otherwise.
  */
 export const request = _.memoize(
-	async (...perms) => {
+	mutex(async (...perms) => {
 		const { permissions, origins } = filterPerms(perms);
+
+		const hasPermissions = await sendMessage('permissions', { operation: 'contains', permissions, origins });
+
+		// already granted, don't ask the user again
+		if (hasPermissions) return;
+
+		const urlStripRe = /((?:\w+\.)+\w+)(?=\/|$)/i;
+
+		await Alert.open(`
+			<p>To create expandos, RES needs permission to access the API(s) at:</p>
+			<br>
+			<p>${origins.map(url => `<code>${urlStripRe.exec(url)[0]}</code>`).join(', ')}</p>
+			<br>
+			<p>Be assured RES does not access any of your information on these domains - it only accesses the API.</p>
```

Comment:
The only thing I'd recommend on that line is to put access/modify any of your information.